### PR TITLE
PCI capabilities and status register support

### PR DESCRIPTION
**Pull request by alesharik**
This PR adds basic PCI capability support (list parsing and MSI capability interface) and status register support

### CODE REVIEW

File: src/capability/msi.rs [hunk 1]
```diff
@@ -0,0 +1,185 @@
+use crate::{capability::PciCapabilityAddress, ConfigRegionAccess};
+use bit_field::BitField;
+use core::convert::TryFrom;
+
+/// Specifies, how much MSI interrupts one device can have.
```
Comment (IsaacWoods) [inline src/capability/msi.rs:5]: Nit: should be "Specifies how many MSI interrupts one device can have"

File: src/capability/msi.rs [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        self.per_vector_masking
+    }
+
+    /// Is device using 64-bit addressing?
+    #[inline]
+    pub fn is_64bit(&self) -> bool {
+        self.is_64bit
+    }
+
+    /// How much interrupts this device has?
```
Comment (IsaacWoods) [inline src/capability/msi.rs:80]: How many

File: src/capability/msi.rs [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    }
+
+    /// Enable or disable MSI capability
+    pub fn set_enabled(&self, enabled: bool, access: &impl ConfigRegionAccess) {
+        let mut reg = unsafe { access.read(self.address.address, self.address.offset) };
+        reg.set_bit(0, enabled);
+        unsafe { access.write(self.address.address, self.address.offset, reg) };
+    }
+
+    /// Set how much interrupts the device will use. If requested count is bigger than supported count,
```
Comment (IsaacWoods) [inline src/capability/msi.rs:99]: "Set how many"

File: src/capability/msi.rs [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+        &self,
+        data: MultipleMessageSupport,
+        access: &impl ConfigRegionAccess,
+    ) {
+        let mut reg = unsafe { access.read(self.address.address, self.address.offset) };
+        reg.set_bits(4..7, (data.min(self.multiple_message_capable)) as u32);
+        unsafe { access.write(self.address.address, self.address.offset, reg) };
+    }
+
+    /// Return how much interrupts the device is using
```
Comment (IsaacWoods) [inline src/capability/msi.rs:111]: "how many"

File: src/capability/mod.rs [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            if self.offset == 0 {
+                return None;
+            }
+            let data = unsafe { self.access.read(self.address, self.offset) };
+            let next_ptr = data.get_bits(8..16);
+            let id = data.get_bits(0..8);
+            let control = data.get_bits(16..32) as u16;
```
Comment (IsaacWoods) [inline src/capability/mod.rs:119]: Hmm so is included with some but not all of the capabilities, and you just know based on the type? Maybe a name like `extension` would be better?

**Review by IsaacWoods — CHANGES_REQUESTED** [review]
Hi, thanks for this contribution! I've not looked at PCI capabilities before but this looks pretty good, left some nits

File: src/capability/mod.rs (hunk 5 above)
Comment (alesharik) [reply src/capability/mod.rs]: It is included with all PCI capabilities, but have different meaning depending on capability type. And yes, `extension` will be a better name than `control`.

**Review by IsaacWoods — APPROVED** [review]
Thanks!

**Pull request merged**